### PR TITLE
rustPlatform: increase build-speed of `checkPhase` for rust-packages

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

<del> :warning: **Caution:** this is more or less an experiment and I'm not sure if that has implications on any leaf-packages since I just tested a few packages like rustup or cargo-audit with it, so this needs some review and testing before becoming mergable.</del>

As noted [later on](https://github.com/nixos/nixpkgs/pull/82342#issuecomment-626018989), I consider this rather usable after several iterations of reviews and tests, however we need more extensive testing before merging this to upstream.

----

When running `cargo test --release`, the artifacts from `buildPhase`
will be reused here. Previously, most of the stuff had to be recompiled
without optimizations.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cole-h** [comment]
Since you note that this needs testing, I tried it on the recently-added `zoxide` package. (If it's useful to know: I run an "other Linux distribution", Arch, as opposed to NixOS.)

After building `cargo`, it proceeded to build the project, followed by tests. With this patch, they completed instantly (because there are none to run). I cannot comfortably speak for the Nix code, but the functionality works fine in this case.

```
running tests
Running cargo cargo test --release --target x86_64-unknown-linux-gnu --frozen --
   Compiling zoxide v0.2.1 (/build/source)
    Finished release [optimized] target(s) in 0.50s
     Running target/x86_64-unknown-linux-gnu/release/deps/zoxide-6631378e48db285a

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by Ma27** [comment]
@cole-h thanks for taking a look at this!

>  If there's anything else I can help with, let me know -- excited to see this land to reduce unnecessary compilation.

Well, this might be pretty time-consuming, but in case someone finds a crate (preferably something used in a NixOS-package) that breaks with this patch it would be fairly helpful (I didn't find any, but I don't consider myself an expert in the rust ecosystem though :) )

File: pkgs/build-support/rust/default.nix
Comment (Mic92) [inline pkgs/build-support/rust/default.nix]: Why was moving this block necessary?

Comment (Mic92) [reply pkgs/build-support/rust/default.nix]: Ah in order to make cargo test remove the code.

Comment (Ma27) [reply pkgs/build-support/rust/default.nix]: That's right. I guess this can be resolved then? : )

**Review by kevincox — APPROVED** [review]
Thanks! I tried (unsuccessfully) to do this a while ago. Looks great!

**Review by cole-h — APPROVED** [review]
FWIW, I built a few packages that I use (`hyperfine`, `alacritty`, and `ripgrep`) with this patch applied and it was all sunshine and rainbows. No more compiling 250+ dependencies twice for a test run!

**Comment by bhipple** [comment]
Looks like a great change in general.  Performance optimizations aside, this is also arguably superior on its own right: we're shipping optimizied binaries built against optimized libraries, but previously we were running the test suite against debug libs. This gets us closer to testing what we're deploying.  The fact that it basically halves our compilations is nice too  :) 

That said, after trying several successfully, I think I found a package that has issues.

@GrahamcOfBorg build zcash

**Comment by Ma27** [comment]
Fix the build issue and also changed the base to `staging` as there's quite much to rebuild :)

File: pkgs/applications/blockchains/zcash/librustzcash/default.nix
Comment (bhipple) [inline pkgs/applications/blockchains/zcash/librustzcash/default.nix]: I see a handful more that need this:

```
$ git grep target/release
target/release
pkgs/applications/blockchains/zcash/librustzcash/default.nix:18:    cp target/release/librustzcash.a $out/lib/
pkgs/applications/misc/alacritty/default.nix:91:    install -D target/release/alacritty $out/bin/alacritty
pkgs/applications/misc/alacritty/default.nix:95:    cp -r target/release/osx/Alacritty.app $out/Applications/Alacritty.app
pkgs/applications/networking/dyndns/cfdyndns/default.nix:22:    cp -p target/release/cfdyndns $out/bin/
pkgs/applications/version-management/git-and-tools/git-ignore/default.nix:27:    cp target/release/build/git-ignore-*/out/git-ignore.1 "$man/man/man1/"
pkgs/development/tools/misc/ycmd/default.nix:73:    TARGET=$out/lib/ycmd/third_party/racerd/target/release
pkgs/development/tools/rust/racerd/default.nix:35:    cp -p target/release/racerd $out/bin/
pkgs/tools/misc/broot/default.nix:26:    OUT_DIR=target/release/build/broot-*/out
pkgs/tools/misc/fd/default.nix:21:    (cd target/release/build/fd-find-*/out
pkgs/tools/misc/lsd/default.nix:17:    install -Dm644 -t $out/share/zsh/site-functions/ target/release/build/lsd-*/out/_lsd
pkgs/tools/misc/lsd/default.nix:18:    install -Dm644 -t $out/share/fish/vendor_completions.d/ target/release/build/lsd-*/out/lsd.fish
pkgs/tools/misc/lsd/default.nix:19:    install -Dm644 -t $out/share/bash-completion/completions/ target/release/build/lsd-*/out/lsd.bash
pkgs/tools/misc/vdirsyncer/default.nix:58:    mkdir -p rust/target/release
pkgs/tools/misc/vdirsyncer/default.nix:59:    ln -s ${native}/lib/libvdirsyncer_rustext* rust/target/release/
pkgs/tools/networking/tox-node/default.nix:27:    install -D target/release/tox-node $out/bin/tox-node
pkgs/tools/text/ripgrep/default.nix:33:    (cd target/release/build/ripgrep-*/out
```

Might be worth flushing them out in a separate PR first, since it's a general improvement to the robustness of packaging unrelated to your specific change here.  Though doing it all in 1 bigger PR is probably fine too.  

I think these packages breaking due to your change is nice side-effect, since it flushes out non-robust package assumptions!

Comment (cole-h) [reply pkgs/applications/blockchains/zcash/librustzcash/default.nix]: I've opened https://github.com/NixOS/nixpkgs/pull/82982 to deal with (most of) these.

**Comment by cole-h** [comment]
While I was messing with https://github.com/NixOS/nixpkgs/pull/82982, I noticed that, when combined with this PR, `ripgrep` would fail to compile because of its `preFixup` -- it wants to `cd $releaseDir/build/ripgrep-*/out`, but there are multiple directories that match that glob (most likely due to the tests)!

I have no idea how this should be worked around, but it's something that should be dealt with before this merges. This worked in the past (I assume) because there was only 1 possible match in the `$releaseDir` due to the tests being run in debug mode.

(I guess I didn't properly test this with `ripgrep` earlier, because it fails similarly on 11.0.2 -- that's a failing on my part and I will try to be more careful about these kinds of assertions in the future.)

**Comment by Ma27** [comment]
> EDIT3: Probably the most annoying side-effect of this change is rg, rg-edd1cb3b1163a7ff, and integration-3cd6ef3798b37d0e get installed to bin/ -- is there any way to reliably detect these hash-appended executables and not copy them?

Not sure yet. In case I have sufficient time, I will start investigating tomorrow :)

**Comment by danieldk** [comment]
I wonder if this is actually a good idea, debug doesn’t just compile with fewer optimizations, it also turns on integer overflow checks and debug assertions (which are used throughout the standard library and crates to do more expensive checks). Doing tests with release builds may miss some bugs which are captured by regular testing using debug builds.

**Comment by kevincox** [comment]
In general I see nix tests as a quick check to ensure that you didn't compile things terribly wrong. In this case I think that integer overflow checks should be caught by the application developers. Debug assertions are more difficult to categorize as catching build/configuration issues but I think that in general it is worth the time savings to turn them off. The job of nix isn't to check that the code is perfect, but mostly to ensure that we built it correctly.

File: pkgs/build-support/rust/default.nix
Comment (Ma27) [inline pkgs/build-support/rust/default.nix]: I see a few issues here: if someone has their own buildPhase, this variable will be undefined (we could check though if it's declared and if not, rerun the `find ...` command). Any further ideas?

**Review by Ma27 — COMMENTED** [review]
> In general I see nix tests as a quick check to ensure that you didn't compile things terribly wrong. In this case I think that integer overflow checks should be caught by the application developers. Debug assertions are more difficult to categorize as catching build/configuration issues but I think that in general it is worth the time savings to turn them off. The job of nix isn't to check that the code is perfect, but mostly to ensure that we built it correctly.

Agreed. See also my comment in discourse: https://discourse.nixos.org/t/speeding-up-rust-app-packaging/6384/8?u=ma27

**Comment by Ma27** [comment]
Rebased onto latest staging to fix the merge-conflict.

**Comment by danieldk** [comment]
> The job of nix isn't to check that the code is perfect, but mostly to ensure that we built it correctly.

Fair enough. I just wanted to mention the downside 😉 .

**Comment by Ma27** [comment]
Rebased onto latest staging and fixed the merge-conflict.

@FRidh any chance to get a hydra-jobset for this (e.g. after 20.03 is released)?

**Comment by Ma27** [comment]
@FRidh @fpletz since 20.03 is released and we probably have some capacity on Hydra, could you add a jobset for this, please? :)

File: pkgs/applications/blockchains/zcash/librustzcash/default.nix
Comment (Ma27) [reply pkgs/applications/blockchains/zcash/librustzcash/default.nix:18]: Just merged the PR of @cole-h (thanks a lot btw!). Will wait until the changes are on staging and then do a rebase to get all of those packages fixed.

**Comment by asymmetric** [comment]
As another data point, [I think](https://salsa.debian.org/rust-team/debcargo/-/blob/cf36bbc8c895211c56a85014778f59fa0218ba0e/src/debian/control.rs#L113) Debian doesn't toggle `--release` when testing (warning: not particularly familiar with the Cargo).

File: pkgs/build-support/rust/default.nix
Comment (cole-h) [reply pkgs/build-support/rust/default.nix:170]: Would it be a bad idea to append this to `postBuild` itself? Something like:

```nix
postBuild = args.postBuild or "" + ''
    bins=$(find $releaseDir \
      -maxdepth 1 \
      -type f \
      -executable ! \( -regex ".*\.\(so.[0-9.]+\|so\|a\|dylib\)" \))
'';
```

This way, if a package defines a `postBuild`, it executes that `postBuild` and then also executes the `find`.

**Comment by Ma27** [comment]
@FRidh any chance to get a Hydra jobset for this (or do you consider this mergeable)? While this seems fairly usable to me now, I'm rather hesitant to merge this to staging without more extensive testing :)

File: pkgs/build-support/rust/default.nix
Comment (Ma27) [reply pkgs/build-support/rust/default.nix:170]: Sorry, forgot to reply here today! The idea sounds reasonable to me, will test it locally and push it if it's fine :)

Comment (Ma27) [reply pkgs/build-support/rust/default.nix:173]: Should be fixed now (unless I misunderstood your suggestion :)).

**Comment by Ma27** [comment]
I'd personally consider this usable now, however I don't have the resources for extensive testing which is why I won't merge this to staging (yet). Are there any further remarks/suggestions/etc to address?

**Review by cole-h — APPROVED** [review]
I built `alacritty`, `ripgrep`, and `zoxide` locally and things work as expected. Even better: no more test binaries (`rg-edd1cb3b1163a7ff` and `integration-3cd6ef3798b37d0e`) show up in the result!

`alacritty` took 6m to build on this branch, and 7m17s on master. An improvement, I must say!

**Comment by FRidh** [comment]
@Ma27 I can't create jobsets but I think @vcunat can

**Comment by vcunat** [comment]
I can do that.  Maybe first base this on a commit that has lots of binaries (and few failures), so you can run Hydra comparisons more easily?  And I assume `x86_64-linux` suffices and nixpkgs type (i.e. without OS tests)?

**Comment by Ma27** [comment]
> @Ma27 I can't create jobsets but I think @vcunat can

Oh, I didn't know. I blindly assumed you had those permissions as well due to your hard work on the staging-workflow, sorry for the noise then!

> I can do that. Maybe first base this on a commit that has lots of binaries (and few failures), so you can run Hydra comparisons more easily?

It seemed pretty fine the last time I rebased onto staging :) 

>  And I assume x86_64-linux suffices and nixpkgs type (i.e. without OS tests)?

Yes. While I obviously tested the functionality of several packages, I wouldn't expect notable differences in the behavior since this is an optimization that applies to the `checkPhase`.

I also think that it's sufficient for now to only build on `x86_64-linux`: at least `aarch64-linux` is actually pretty well-tested since I originally developed this with the community-builder since it has way more resources than my personal build infra :)

**Comment by vcunat** [comment]
https://hydra.nixos.org/eval/1586778?compare=1586582

I wanted a good base for comparison while not breaking the diff here on GitHub, so I retargeted it to staging-next.

**Comment by Ma27** [comment]
Pushed some fixes for packages from the `rust`-tarball (like `rustfmt`) where you have to move into a subdir for building/testing. Eval is already running.

File: pkgs/build-support/rust/default.nix
Comment (Ma27) [reply pkgs/build-support/rust/default.nix:194]: I reverted this change again. This would break packages like `uutils-coreutils` with custom `build`/`installPhase` where you'd have to set `postBuild = ""` otherwise.

Comment (cole-h) [reply pkgs/build-support/rust/default.nix:194]: Unfortunate. What if you did the above suggestion but for `buildPhase` instead of `postBuild`? Like

```nix
buildPhase = with builtins; args.buildPhase or ''
      # the stuff that's already there
    '' + ''
      bins=$(find $releaseDir \
        -maxdepth 1 \
        -type f \
        -executable ! \( -regex ".*\.\(so.[0-9.]+\|so\|a\|dylib\)" \))
    '';
```

Do you think that would work for `uutils-coreutils` (and the others having problems)?

**Comment by vcunat** [comment]
Hydra now shows only three repeatable regressions: https://hydra.nixos.org/eval/1587152?compare=1586582#tabs-now-fail

**Comment by Ma27** [comment]
Just pushed the fixes for the three remaining packages.

While the changes seem usable now, the following things need to be done before merging this:

* I think that we need some more feedback from the Rust-users in the community. I can write a Discourse Post for this (unless someone else is faster :p).
* We need at least something in the release-notes and probably some more docs.
* This is now a breaking change since the normalization of the paths in `target/` is done in `installPhase` and I'm surprised how many tests rely on the structure in `target/`. I need to think about this (and the suggestions of @cole-h about this) and probably make this less breaking.

**Review by Br1ght0ne — APPROVED** [review]

**Review by stigtsp — APPROVED** [review]
`ripasso-cursive` looks good to me

**Comment by Ma27** [comment]
So, as I didn't receive any negative feedback so far I guess that the overall change is rather mature. I'll write some docs this weekend, after that, this PR should be mergable from my PoV.

**Comment by Ma27** [comment]
Also updated the docs accordingly. Would be great to receive a bit more feedback, later I'd rebase onto a recent staging and after a last Hydra eval it should be good to go, at least from my PoV.

File: pkgs/applications/editors/neovim/gnvim/default.nix
Comment (Mic92) [inline pkgs/applications/editors/neovim/gnvim/default.nix:38]:
```suggestion
  buildTargets = [ "build" ];
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:83]: I think "to actually test the artifacts that will be used" is a bit unnecessary. What else would `cargo test` be testing?

```suggestion
When using `buildRustPackage`, the `checkPhase` is enabled by default and runs
`cargo test` on the package. In order to prevent compiling the source twice, the
tests will be ran in `release` mode by default.
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:86]: "in some cases" followed by "in that case" in the next sentence is too repetitive.

```suggestion
However, in some cases, the test-suite of a package doesn't work properly in the
`release` mode. For these situations, the mode for `checkPhase` can be changed like so:
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:101]: Nits.

```suggestion
#### Tests relying on the structure of the `target/` directory

Some tests may rely on the structure of the `target/` directory. Those tests
are likely to fail because we use `cargo --target` during the build. This means that
the artifacts
[are stored in `target/<architecture>/release/`](https://doc.rust-lang.org/cargo/guide/build-cache.html),
rather than `target/release/`.
```

File: pkgs/applications/editors/neovim/gnvim/default.nix
Comment (Ma27) [reply pkgs/applications/editors/neovim/gnvim/default.nix:38]: If the `buildPhase` isn't changed, the `buildPhase` from `buildRustPackage` would've been used, right?

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:108]:
```suggestion
In some instances, it may be necessary to disable testing with `doCheck = false;`.
Acceptable reasons for doing so include the following:
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:114]:
```suggestion
* If no tests exist -- the `checkPhase` should be explicitly disabled to skip
  unnecessary build-steps to speed-up the build.

* If tests are highly impure (e.g. due to network usage).
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:120]:
```suggestion
There will obviously be some corner-cases not listed above where it's sensible to disable tests.
The above are just guidelines, and exceptions may be granted on a case-by-case basis.

However, please check if it's possible to disable a problematic subset of the
test suite and leave a comment explaining your reasoning.
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:125]:
```suggestion
### Building a package in `debug` mode

By default, `buildRustPackage` will use `release` mode for builds. If a package
should be built in `debug` mode, it can be configured like so:
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:134]:
```suggestion
In this scenario, the `checkPhase` will be ran in `debug` mode as well.
```

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:142]:
```suggestion
Some packages may use custom scripts for building/installing, e.g. with a `Makefile`.
In these cases, it's recommended to override the `buildPhase`/`installPhase`/`checkPhase`.

Otherwise, some steps may fail because of the modified directory structure of `target/`.
```

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (cole-h) [inline nixos/doc/manual/release-notes/rl-2009.xml:363]:
```suggestion
     Packages built using <literal>buildRustPackage</literal> now use <literal>release</literal>
```

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (cole-h) [inline nixos/doc/manual/release-notes/rl-2009.xml:370]: I don't see why `rust` needs to be a literal, rather than just using "Rust".

```suggestion
     Please note that Rust packages utilizing a custom build/install procedure
     (e.g. by using a <filename>Makefile</filename>) or test suites that rely on the
     structure of the <filename>target/</filename> directory may break due to those assumptions.
     For further information, please read the Rust section in the Nixpkgs manual.
```

**Review by cole-h — CHANGES_REQUESTED** [review]
Some wording suggestions (a lot of them, really).

Reflow as necessary.

**Review by Br1ght0ne — APPROVED** [review]
Besides wording suggestions by @cole-h, LGTM as far as I'm concerned.

Tested on some of the packages I maintain:
- `bandwhich`
- `emplace`
- `nushell`

All of those seem to build and pass tests.

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [inline doc/languages-frameworks/rust.section.md:82]:
```suggestion
sources twice and to actually test the artifacts that will be used at runtime, 
```

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [inline doc/languages-frameworks/rust.section.md:83]:
```suggestion
the tests will be ran in the `release` mode by default.
```

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [inline doc/languages-frameworks/rust.section.md:101]:
```suggestion
rather than in `target/release/`.
```

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [inline doc/languages-frameworks/rust.section.md:103]:
```suggestion
This can only be worked around by patching the affected tests accordingly.
```

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [reply doc/languages-frameworks/rust.section.md:108]: @cole-h I would add *altogether*:

> In some instances, it may be necessary to disable testing altogether (with `doCheck = false;`)

File: doc/languages-frameworks/rust.section.md
Comment (asymmetric) [inline doc/languages-frameworks/rust.section.md:111]:
```suggestion
  unnecessary build steps to speed up the build.
```

File: doc/languages-frameworks/rust.section.md
Comment (Ma27) [reply doc/languages-frameworks/rust.section.md:83]: > I think "to actually test the artifacts that will be used" is a bit unnecessary. What else would cargo test be testing? 

The build artifacts from the `debug` mode (i.e. without optimizations etc.).

File: pkgs/applications/editors/neovim/gnvim/default.nix
Comment (Mic92) [reply pkgs/applications/editors/neovim/gnvim/default.nix:38]: Right.

**Comment by Ma27** [comment]
Docs have been fixed, thanks @cole-h and @asymmetric !

I'd love to see a final review, after that I'd consider this mergable (will leave a comment, then). /cc @FRidh @vcunat

**Comment by bhipple** [comment]
Sadly I don't have time for a full review/test, but I'll at least comment to say I'm morally in favor of this change.

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:85]: I think it should be noted explicitly that using `checkType = "debug";` will compile two times -- in release mode for the actual package, and in debug mode for testing (as was happening before this change).

e.g. "NOTE: This will cause two compilations -- one in `release` mode for the package itself, and one in `debug` mode for the tests." or something.

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:120]:
```suggestion
### Building a package in `debug` mode
```

**Review by cole-h — APPROVED** [review]
A few nits, and the rest LGTM. Thanks for your work on this! Can't wait to compile packages like Alacritty only once when testing ;-)

**Comment by Ma27** [comment]
@cole-h fixed :)

File: doc/languages-frameworks/rust.section.md
Comment (cole-h) [inline doc/languages-frameworks/rust.section.md:97]: One last nit (we dropped "the" from "the `....` mode" everywhere else, should do it here as well):

```suggestion
Please note that the code will be compiled twice here: once in `release` mode
for the `buildPhase`, and again in `debug` mode for the `checkPhase`.
```

**Pull request merged**
